File: app/src/main/java/com/rowland/delivery/features/dash/presentation/fragments/OrderItemFragment.java
package com.rowland.delivery.features.dash.presentation.fragments;


import android.Manifest;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.dekoservidoni.omfm.OneMoreFabMenu;
import com.jude.easyrecyclerview.EasyRecyclerView;
import com.rowland.delivery.features.dash.di.modules.order.OrderModule;
import com.rowland.delivery.features.dash.domain.models.order.OrderData;
import com.rowland.delivery.features.dash.presentation.activities.DashActivity;
import com.rowland.delivery.features.dash.presentation.adapters.OrderItemAdapter;
import com.rowland.delivery.features.dash.presentation.viewmodels.order.OrderViewModel;
import com.rowland.delivery.merchant.R;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

/**
 * Created by Rowland on 5/7/2018.
 */
public class OrderItemFragment extends Fragment {

    private Unbinder unbinder;
    private OrderViewModel orderViewModel;
    private OrderData orderData;
    private OrderItemAdapter adapter;
    private static final int CALL_PERMISSION = 100;

    @BindView(R.id.fab)
    OneMoreFabMenu fab;

    @BindView(R.id.main_toolbar_orders)
    Toolbar toolbar;

    @BindView(R.id.customer_name)
    TextView customer_name;

    @BindView(R.id.status)
    TextView status;

    @BindView(R.id.phone)
    TextView phone;

    @BindView(R.id.ordered_items)
    EasyRecyclerView recyclerViewManager;

    @BindView(R.id.email)
    TextView email;

    @BindView(R.id.address)
    TextView address;

    @BindView(R.id.order_ref)
    TextView order_ref;

    @BindView(R.id.all_items)
    TextView all_items;

    @BindView(R.id.total_item_price)
    TextView total_item_price;

    @BindView(R.id.delivery_fee)
    TextView delivery_fee;

    @BindView(R.id.total_price)
    TextView total_price;


    @Inject
    @Named("order")
    ViewModelProvider.Factory orderFactory;

    public OrderItemFragment() {
        // Required empty public constructor
    }

    public static OrderItemFragment newInstance(Bundle args) {
        OrderItemFragment frag = new OrderItemFragment();
        frag.setArguments(args);
        return frag;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        orderViewModel = ViewModelProviders.of(getActivity(), orderFactory).get(OrderViewModel.class);
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_order_item, container, false);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        ((DashActivity) getActivity()).getDashComponent()
                .getOrderComponent(new OrderModule())
                .inject(this);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        unbinder = ButterKnife.bind(this, view);

        if (toolbar != null) {
            ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);
        }

        adapter = new OrderItemAdapter();
        recyclerViewManager.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerViewManager.setItemAnimator(new DefaultItemAnimator());
        recyclerViewManager.setAdapterWithProgress(adapter);

        fab.setOptionsClick(id -> {
            switch (id) {
                case R.id.action_pending:
                    changeStatus("pending");
                    break;
                case R.id.action_complete:
                    changeStatus("completed");
                    break;
                case R.id.action_in_progress:
                    changeStatus("in_progress");
                    break;
                case R.id.action_canceled:
                    changeStatus("cancelled");
            }
            fab.performClick();
        });

        orderViewModel.getSelectedOrderData()
                .observe(this, orderData -> {
                    this.orderData = orderData;
                    setupData();
                });
    }

    @Override
    public void onDetach() {
        super.onDetach();
        unbinder.unbind();
    }

    @OnClick(R.id.action_call_btn)
    public void makeCall() {
        if (ActivityCompat.checkSelfPermission(getActivity(), Manifest.permission.CALL_PHONE) == PackageManager.PERMISSION_GRANTED) {
            Intent callIntent = new Intent(Intent.ACTION_CALL);
            callIntent.setData(Uri.parse(String.format("tel:%s", orderData.getPhone())));
            startActivity(callIntent);
        } else {
            ActivityCompat.requestPermissions(getActivity(), new String[]{Manifest.permission.CALL_PHONE}, CALL_PERMISSION);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode == CALL_PERMISSION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                makeCall();
            } else {
                Toast.makeText(getActivity(), "Please grant permission to call", Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void changeTextStatus(String s) {
        status.setText(s);
        switch (s) {
            case "active":
                status.setTextColor(Color.parseColor("#378E3D"));
                break;
            case "delivered":
                status.setTextColor(Color.parseColor("#FF6D00"));
                break;
            case "failed":
                status.setTextColor(Color.parseColor("#B94464"));
                break;
            case "missed":
                status.setTextColor(Color.parseColor("#B94464"));
                break;
            case "completed":
                status.setTextColor(Color.parseColor("#AAAAAA"));
                break;
            case "cancelled":
                status.setTextColor(Color.parseColor("#B94464"));
                break;
            case "in_progress":
                status.setTextColor(Color.parseColor("#378E3D"));
                break;
            case "pending":
                status.setTextColor(Color.parseColor("#B94464"));
                break;
            default:
                status.setTextColor(Color.parseColor("#AAAAAA"));
        }
    }

    private final void changeStatus(String status) {
        orderViewModel.updateOrderStatus(status)
                .subscribe(() -> Toast.makeText(getActivity(), "Order Status Updated Successfull", Toast.LENGTH_SHORT).show()
                        , throwable -> Toast.makeText(getActivity(), "Update Failed", Toast.LENGTH_SHORT).show());
    }


    private void setupData() {
        customer_name.setText(orderData.getName());
        changeTextStatus(orderData.getStatus());

        phone.setText(orderData.getPhone());
        email.setText("Email not provided");

        if (orderData.getAddress().isEmpty()) {
            address.setText("Address not provided");
        } else {
            address.setText(orderData.getAddress());
        }

        order_ref.setText(orderData.getOrderRef());
        all_items.setText("items (" + orderData.getOrderItemsQuantity() + ")");
        total_item_price.setText("Ksh " + orderData.getOrderSubTotal());
        delivery_fee.setText("Ksh " + orderData.getDeliveryFee());
        total_price.setText("Ksh " + orderData.getOrderTotal());

        adapter.setList(orderData.getItems());
        adapter.notifyDataSetChanged();
    }
}

File: ui/src/main/java/com/rowland/delivery/merchant/features/dash/fragments/OrderItemFragment.kt
/*
 * Copyright (c) Otieno Rowland,  2021. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rowland.delivery.merchant.features.dash.fragments

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import com.dekoservidoni.omfm.OneMoreFabMenu
import com.rowland.delivery.merchant.R
import com.rowland.delivery.merchant.R.string
import com.rowland.delivery.merchant.databinding.FragmentOrderItemBinding
import com.rowland.delivery.merchant.features.dash.adapters.OrderItemAdapter
import com.rowland.delivery.presentation.viewmodels.order.OrderViewModel
import dagger.hilt.android.AndroidEntryPoint
import java.util.HashMap
import javax.inject.Inject

/**
 * Created by Rowland on 5/7/2018.
 */

@AndroidEntryPoint
class OrderItemFragment : Fragment(), OneMoreFabMenu.OptionsClick {

    private val orderViewModel: OrderViewModel by viewModels()
    private val args by navArgs<OrderItemFragmentArgs>()

    @Inject
    lateinit var adapter: OrderItemAdapter

    private var _binding: FragmentOrderItemBinding? = null
    private val binding get() = _binding!!

    companion object {

        private const val CALL_PERMISSION = 100

        fun newInstance(args: Bundle?): OrderItemFragment {
            val frag = OrderItemFragment()
            frag.arguments = args
            return frag
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentOrderItemBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.ordereditemRecyclerview.setLayoutManager(LinearLayoutManager(activity))
        binding.ordereditemRecyclerview.setItemAnimator(DefaultItemAnimator())
        binding.ordereditemRecyclerview.setAdapterWithProgress(adapter)

        setupData()

        binding.actionCallBtn.setOnClickListener { makeCall() }
        binding.orderitemFab.setOptionsClick(this)
    }

    override fun onOptionClick(optionId: Int?) {
        when (optionId!!) {
            R.id.action_pending -> changeStatus("pending")
            R.id.action_complete -> changeStatus("completed")
            R.id.action_in_progress -> changeStatus("in_progress")
            R.id.action_canceled -> changeStatus("cancelled")
        }
        binding.orderitemFab.performClick()
    }

    private fun makeCall() {
        if (ActivityCompat.checkSelfPermission(
                requireActivity(),
                Manifest.permission.CALL_PHONE
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            val callIntent = Intent(Intent.ACTION_CALL)
            callIntent.data = Uri.parse(String.format("tel:%s", args.orderDataItem.phone))
            startActivity(callIntent)
        } else {
            ActivityCompat.requestPermissions(
                requireActivity(),
                arrayOf(Manifest.permission.CALL_PHONE),
                CALL_PERMISSION
            )
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == CALL_PERMISSION) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                makeCall()
            } else {
                Toast.makeText(activity, getString(string.permision_to_call_msg), Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun changeTextStatus(s: String) {
        val status = binding.status
        status.text = s
        when (s) {
            "active" -> status.setTextColor(Color.parseColor("#378E3D"))
            "delivered" -> status.setTextColor(Color.parseColor("#FF6D00"))
            "failed" -> status.setTextColor(Color.parseColor("#B94464"))
            "missed" -> status.setTextColor(Color.parseColor("#B94464"))
            "completed" -> status.setTextColor(Color.parseColor("#AAAAAA"))
            "cancelled" -> status.setTextColor(Color.parseColor("#B94464"))
            "in_progress" -> status.setTextColor(Color.parseColor("#378E3D"))
            "pending" -> status.setTextColor(Color.parseColor("#B94464"))
            else -> status.setTextColor(Color.parseColor("#AAAAAA"))
        }
    }

    @SuppressLint("CheckResult")
    private fun changeStatus(status: String) {
        val orderUpdateFields = HashMap<String, Any>()
        orderUpdateFields["status"] = status
        orderViewModel.updateOrder(orderUpdateFields, args.orderDataItem.firestoreUid!!)
            .subscribe({
                Toast.makeText(activity, getString(string.order_status_update_success), Toast.LENGTH_SHORT).show()
            }) {
                Toast.makeText(
                    activity,
                    getString(
                        string.order_status_update_failed
                    ),
                    Toast.LENGTH_SHORT
                ).show()
            }
    }

    private fun setupData() {
        (activity as AppCompatActivity?)?.supportActionBar?.let { it.subtitle = args.orderDataItem.orderRef }
        binding.customerName.text = args.orderDataItem.name
        changeTextStatus(args.orderDataItem.status!!)

        binding.phone.text = args.orderDataItem.phone
        binding.email.text = getString(string.msg_email_unprovided)

        if (args.orderDataItem.address!!.isEmpty()) {
            binding.address.text = getString(string.msg_address_unprovided)
        } else {
            binding.address.text = args.orderDataItem.address
        }

        binding.allItems.text = getString(string.items_label_with_value_num, args.orderDataItem.orderItemsQuantity)
        binding.totalItemPrice.text = getString(string.ksh_label_with_value_str, args.orderDataItem.orderSubTotal)
        binding.deliveryFee.text = getString(string.ksh_label_with_value_num, args.orderDataItem.deliveryFee!!)
        binding.totalPrice.text = getString(string.ksh_label_with_value_num, args.orderDataItem.orderTotal!!)

        adapter.setList(args.orderDataItem.items!!)
        adapter.notifyDataSetChanged()
    }
}
